File: web/lib/api/modules/user_transfer.ts
import { apiGet, apiPost } from "@/lib/api/core/http";

export type UserTransferRecipient = {
  user_id: number;
  email_mask: string;
  nickname?: string | null;
  avatar_url?: string | null;
  can_transfer: boolean;
};

export type CreateUserTransferPayload = {
  request_id: string;
  recipient_email: string;
  symbol: string;
  amount: string;
  remark?: string;
};

export type UserTransferRecord = {
  id: number;
  transfer_no: string;
  request_id: string;
  direction: "in" | "out";
  counterparty_user_id: number;
  counterparty_nickname?: string | null;
  recipient_nickname?: string | null;
  recipient_email_mask: string;
  symbol: string;
  from_account: "funding";
  to_account: "funding";
  amount: string;
  fee_amount: string;
  net_amount: string;
  status: string;
  sender_available_before: string;
  sender_available_after: string;
  receiver_available_before: string;
  receiver_available_after: string;
  remark?: string | null;
  created_at: string;
};

export type CreateUserTransferResponse = {
  record: UserTransferRecord;
};

export type GetUserTransferRecordsParams = {
  direction?: "all" | "in" | "out";
  page?: number;
  page_size?: number;
  symbol?: string;
};

export type GetUserTransferRecordsResponse = {
  items: UserTransferRecord[];
  total: number;
  page: number;
  page_size: number;
};

const UserTransferAPI = {
  async resolveRecipient(email: string): Promise<UserTransferRecipient> {
    const qs = new URLSearchParams();
    qs.set("email", email.trim());
    return apiGet<UserTransferRecipient>(`/user-transfer/recipient/resolve?${qs.toString()}`);
  },

  async createTransfer(payload: CreateUserTransferPayload): Promise<CreateUserTransferResponse> {
    return apiPost<CreateUserTransferResponse, CreateUserTransferPayload>("/user-transfer", payload);
  },

  async getRecords(params: GetUserTransferRecordsParams = {}): Promise<GetUserTransferRecordsResponse> {
    const qs = new URLSearchParams();
    Object.entries(params).forEach(([key, value]) => {
      if (value === undefined || value === null || value === "") return;
      qs.set(key, String(value));
    });
    const suffix = qs.toString();
    return apiGet<GetUserTransferRecordsResponse>(
      suffix ? `/user-transfer/records?${suffix}` : "/user-transfer/records"
    );
  },
};

export default UserTransferAPI;

File: backend/app/schemas/user_transfer.py
from __future__ import annotations

from decimal import Decimal
from typing import List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


class UserTransferRecipientData(BaseModel):
    user_id: int
    email_mask: str
    nickname: Optional[str] = None
    avatar_url: Optional[str] = None
    can_transfer: bool


class UserTransferRequest(BaseModel):
    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "request_id": "b43bbf1b-54d1-4f71-8e70-f8d43d56f7f6",
                "recipient_email": "user@example.com",
                "symbol": "USDT",
                "amount": "10",
                "remark": "optional",
            }
        }
    )

    request_id: str = Field(..., min_length=1, max_length=64)
    recipient_email: str = Field(..., min_length=3, max_length=191)
    symbol: str = Field(..., min_length=1, max_length=32)
    amount: Decimal = Field(...)
    remark: Optional[str] = Field(None, max_length=255)


class UserTransferRecordItem(BaseModel):
    id: int
    transfer_no: str
    request_id: str
    direction: Literal["in", "out"]
    counterparty_user_id: int
    counterparty_nickname: Optional[str] = None
    recipient_nickname: Optional[str] = None
    recipient_email_mask: str
    symbol: str
    from_account: Literal["funding"]
    to_account: Literal["funding"]
    amount: str
    fee_amount: str
    net_amount: str
    status: str
    sender_available_before: str
    sender_available_after: str
    receiver_available_before: str
    receiver_available_after: str
    remark: Optional[str] = None
    created_at: str


class UserTransferSubmitData(BaseModel):
    record: UserTransferRecordItem


class UserTransferRecordsData(BaseModel):
    items: List[UserTransferRecordItem]
    total: int
    page: int
    page_size: int
